### PR TITLE
[Feature] Support Class Aware Sampler

### PR DESCRIPTION
**Pull request by BIGWangYuDong**
This PR is based on https://github.com/open-mmlab/mmdetection/pull/6331.

More details will be updated when this PR is fully prepared

### CODE REVIEW

File: configs/openimages/README.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+# Open Images Dataset
+<!-- [DATASET] -->
+
+## Abstract
+
+<!-- [ABSTRACT] -->
+#### Open Images v6
+
+[Open Images](https://storage.googleapis.com/openimages/web/index.html) is a dataset of ~9M images annotated with image-level labels,
```
Comment (jbwang1997) [inline configs/openimages/README.md:9]: Need to use the new format of README.md

Comment (BIGWangYuDong) [reply configs/openimages/README.md:9]: OK, I'll fix it in https://github.com/open-mmlab/mmdetection/pull/6331.

File: mmdet/datasets/custom.py [hunk 2]
```diff
@@ -285,6 +285,27 @@ def get_classes(cls, classes=None):
...
+    def get_label_dict(self):
+        """Get per-label image list in the current dataset, which will be used
+        in :class:`ClassAwareSampler`.
+
+        Returns:
+            defaultdict[list]: A defaultdict of per-label image list,
+            the item of the defaultdict indicates a label index,
+            corresponds to the image index that contains the label.
+        """
+        label_dict = defaultdict(int)
```
Comment (jbwang1997) [inline mmdet/datasets/custom.py:297]: Can we directly set `label_dict = defaultdict(list)`.

File: mmdet/datasets/custom.py [hunk 3]
```diff
@@ -285,6 +285,27 @@ def get_classes(cls, classes=None):
...
+        if self.CLASSES is None:
+            raise ValueError('CLASSES can not be None')
+        for i in range(len(self.CLASSES)):
+            label_dict[i] = []
```
Comment (jbwang1997) [inline mmdet/datasets/custom.py:301]: If we set `label_dict = defaultdict(list)`, this part can be ignored.

File: mmdet/datasets/pipelines/loading.py [hunk 4]
```diff
@@ -246,11 +251,24 @@ def _load_bboxes(self, results):
         ann_info = results['ann_info']
         results['gt_bboxes'] = ann_info['bboxes'].copy()
 
+        if self.denorm_bbox:
+            h, w = results['img_shape'][:2]
+            bbox_num = results['gt_bboxes'].shape[0]
+            if bbox_num != 0:
+                results['gt_bboxes'][:, 0::2] *= w
+                results['gt_bboxes'][:, 1::2] *= h
+            results['gt_bboxes'] = results['gt_bboxes'].astype(np.float32)
```
Comment (jbwang1997) [inline mmdet/datasets/pipelines/loading.py:260]: Do we also need to denormalize `gt_bboxes_ignore`

**Comment by jbwang1997** [comment]
We need to add `openimages/metafile.yml` into [model-index.yml](https://github.com/open-mmlab/mmdetection/blob/master/model-index.yml).

File: configs/openimages/README.md [hunk 5]
```diff
@@ -1,6 +1,9 @@
 # Open Images Dataset
-<!-- [DATASET] -->
 
+> [Open Images Dataset](https://arxiv.org/abs/1811.00982)
+
+<!-- [DATASET] -->
+<!-- [ALGORITHM] -->
```
Comment (BIGWangYuDong) [inline configs/openimages/README.md:6]: will get error when using pre-commit
```
check algorithm readme...................................................Failed
- hook id: check-algo-readme
- exit code: 1

Failed to find "<!-- [ALGORITHM] -->" flag from readme, please check configs/openimages/README.md again.
```

Comment (jbwang1997) [reply configs/openimages/README.md:6]: Only algorithm class has Collections in metafile.

**Comment by BIGWangYuDong** [comment]
kindly ping @jbwang1997 have a review

File: configs/openimages/README.md (hunk 5 above)
Comment (jbwang1997) [reply configs/openimages/README.md:6]: Do we need to delete  here?

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+    def __iter__(self):
+        # deterministically shuffle based on epoch
+        g = torch.Generator()
+        g.manual_seed(self.epoch + self.seed)
+
+        # initialize label list
+        label_iter_list = RandomCycleIter(self.class_index, generator=g)
+        # initialize each per-label image list
+        data_iter_dict = defaultdict(int)
```
Comment (jbwang1997) [inline mmdet/datasets/samplers/class_aware_sampler.py:87]: `defaultdict` is unnecessary, can replace it with `dict`.

File: mmdet/datasets/openimages.py [hunk 7]
```diff
@@ -66,6 +69,7 @@ def __init__(self,
                  meta_file='',
                  filter_labels=True,
                  load_image_level_labels=True,
+                 num_sample_class=1,
```
Comment (jbwang1997) [inline mmdet/datasets/openimages.py:72]: Can we put num_sample_class in out of dataset?

Comment (jbwang1997) [reply mmdet/datasets/openimages.py:72]: We can get this argument in **kwargs of `build_dataloader`.

**Comment by BIGWangYuDong** [comment]
Move to #7436

**Pull request closed without merging**